File: backend/app/schemas/subscription.py
"""Pydantic models for subscription-related schemas."""
from datetime import datetime
from typing import List, Optional, Dict, Any
from enum import Enum
from pydantic import BaseModel, Field, validator

class SubscriptionTier(str, Enum):
    FREE = "free"
    PRO = "pro"
    BUSINESS = "business"
    ADMIN = "admin"

class SubscriptionStatus(str, Enum):
    ACTIVE = "active"
    CANCELED = "canceled"
    PAST_DUE = "past_due"
    UNPAID = "unpaid"
    TRIALING = "trialing"
    INCOMPLETE = "incomplete"
    INCOMPLETE_EXPIRED = "incomplete_expired"
    PAUSED = "paused"

class SubscriptionBase(BaseModel):
    tier: SubscriptionTier = Field(..., description="Subscription tier")
    status: SubscriptionStatus = Field(..., description="Subscription status")
    current_period_start: Optional[datetime] = Field(None, description="Start of the current billing period")
    current_period_end: Optional[datetime] = Field(None, description="End of the current billing period")
    cancel_at_period_end: bool = Field(False, description="Whether the subscription will be canceled at the end of the billing period")

class SubscriptionCreate(SubscriptionBase):
    payment_method_id: Optional[str] = Field(None, description="ID of the payment method to use for this subscription")
    trial_period_days: Optional[int] = Field(None, description="Number of days for the trial period")

class SubscriptionUpdate(BaseModel):
    tier: Optional[SubscriptionTier] = Field(None, description="New subscription tier")
    status: Optional[SubscriptionStatus] = Field(None, description="New subscription status")
    cancel_at_period_end: Optional[bool] = Field(None, description="Whether to cancel the subscription at the end of the billing period")
    payment_method_id: Optional[str] = Field(None, description="New payment method ID to use for this subscription")

class SubscriptionInDB(SubscriptionBase):
    id: int
    user_id: int
    created_at: datetime
    updated_at: datetime
    metadata: Dict[str, Any] = {}

    class Config:
        orm_mode = True

class InvoiceBase(BaseModel):
    amount: int = Field(..., description="Amount in the smallest currency unit (e.g., cents)")
    currency: str = Field("usd", description="Three-letter ISO currency code")
    status: str = Field(..., description="Invoice status")
    paid: bool = Field(False, description="Whether the invoice has been paid")
    receipt_url: Optional[str] = Field(None, description="URL to view the invoice")

class InvoiceCreate(InvoiceBase):
    subscription_id: int = Field(..., description="ID of the subscription this invoice is for")
    payment_intent_id: Optional[str] = Field(None, description="ID of the payment intent for this invoice")

class InvoiceInDB(InvoiceBase):
    id: int
    subscription_id: int
    created_at: datetime
    paid_at: Optional[datetime] = None

    class Config:
        orm_mode = True

class SubscriptionWithInvoices(SubscriptionInDB):
    invoices: List[InvoiceInDB] = []

class SubscriptionPlan(BaseModel):
    id: str = Field(..., description="Plan ID")
    name: str = Field(..., description="Plan name")
    description: str = Field(..., description="Plan description")
    price: int = Field(..., description="Price in the smallest currency unit (e.g., cents)")
    currency: str = Field("usd", description="Three-letter ISO currency code")
    interval: str = Field(..., description="Billing interval (e.g., 'month' or 'year')")
    features: List[str] = Field(..., description="List of features included in this plan")
    is_popular: bool = Field(False, description="Whether this is a featured/popular plan")

class SubscriptionFeatures(BaseModel):
    tier: SubscriptionTier = Field(..., description="Subscription tier")
    features: Dict[str, Any] = Field(..., description="Dictionary of features and their values")

class SubscriptionStatusResponse(BaseModel):
    has_access: bool = Field(..., description="Whether the user has access to the requested feature")
    message: str = Field(..., description="Explanation of the access status")
    subscription: Optional[SubscriptionInDB] = Field(None, description="User's subscription details")
    features: Dict[str, Any] = Field(default_factory=dict, description="Available features for the user's subscription")

File: temp_frontend/src/services/api/subscription.ts
import axios from 'axios';

// Using process.env for Vite environment variables
const API_BASE_URL = (import.meta as any).env.VITE_API_URL || 'http://localhost:3001/api';

export interface SubscriptionTier {
  id: string;
  name: string;
  description: string;
  price: number;
  interval: 'month' | 'year';
  features: string[];
  isPopular?: boolean;
  limits: {
    transcriptionMinutes: number;
    aiGenerations: number;
    storageMB: number;
  };
}

export interface Subscription {
  id: string;
  userId: string;
  tierId: string;
  status: 'active' | 'canceled' | 'past_due' | 'unpaid';
  currentPeriodStart: string;
  currentPeriodEnd: string;
  cancelAtPeriodEnd: boolean;
  usage: {
    transcriptionMinutes: {
      used: number;
      total: number;
    };
    aiGenerations: {
      used: number;
      total: number;
    };
    storageMB: {
      used: number;
      total: number;
    };
  };
}

export const subscriptionApi = {
  // Get all available subscription tiers
  async getSubscriptionTiers(): Promise<SubscriptionTier[]> {
    try {
      const response = await axios.get<SubscriptionTier[]>(
        `${API_BASE_URL}/subscription/tiers`,
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to fetch subscription tiers:', error);
      throw error;
    }
  },

  // Get current user's subscription
  async getCurrentSubscription(): Promise<Subscription | null> {
    try {
      const response = await axios.get<Subscription>(
        `${API_BASE_URL}/subscription`,
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response?.status === 404) {
        return null; // No active subscription
      }
      console.error('Failed to fetch current subscription:', error);
      throw error;
    }
  },

  // Create a checkout session for a subscription
  async createCheckoutSession(tierId: string): Promise<{ url: string }> {
    try {
      const response = await axios.post<{ url: string }>(
        `${API_BASE_URL}/subscription/checkout`,
        { tierId },
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to create checkout session:', error);
      throw error;
    }
  },

  // Update subscription (upgrade/downgrade)
  async updateSubscription(tierId: string): Promise<Subscription> {
    try {
      const response = await axios.put<Subscription>(
        `${API_BASE_URL}/subscription`,
        { tierId },
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to update subscription:', error);
      throw error;
    }
  },

  // Cancel subscription
  async cancelSubscription(): Promise<Subscription> {
    try {
      const response = await axios.post<Subscription>(
        `${API_BASE_URL}/subscription/cancel`,
        {},
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to cancel subscription:', error);
      throw error;
    }
  },

  // Get subscription usage
  async getUsage(): Promise<Subscription['usage']> {
    try {
      const response = await axios.get<Subscription['usage']>(
        `${API_BASE_URL}/subscription/usage`,
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to fetch subscription usage:', error);
      throw error;
    }
  },

  // Update payment method
  async updatePaymentMethod(paymentMethodId: string): Promise<void> {
    try {
      await axios.post(
        `${API_BASE_URL}/subscription/payment-method`,
        { paymentMethodId },
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
    } catch (error) {
      console.error('Failed to update payment method:', error);
      throw error;
    }
  },

  // Get billing history
  async getBillingHistory(): Promise<Array<{
    id: string;
    amount: number;
    currency: string;
    date: string;
    status: 'paid' | 'pending' | 'failed';
    receiptUrl?: string;
  }>> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/subscription/invoices`,
        {
          headers: {
            'Content-Type': 'application/json',
          },
          withCredentials: true,
        }
      );
      return response.data;
    } catch (error) {
      console.error('Failed to fetch billing history:', error);
      throw error;
    }
  },
};

export default subscriptionApi;
